pkg/catalogv2/git/git.go

```diff
@@ -1,480 +1,440 @@
+// and configurates the git repo communications accordingly
+func (r *repository) setRepoCredentials() error {
+	if r.secret == nil {
+		return nil
 	}
 
-	if err := g.reset("HEAD"); err != nil {
-		return "", err
-	}
+	var username string
+	var password string
 
-	return g.currentCommit()
-}
+	switch r.secret.Type {
+	case corev1.SecretTypeBasicAuth: // BASIC HTTP(S) AUTHENTICATION
+		// get the credentials setted in kubernetes
+		username = string(r.secret.Data[corev1.BasicAuthUsernameKey])
+		password = string(r.secret.Data[corev1.BasicAuthPasswordKey])
+		if len(password) == 0 || len(username) == 0 {
+			return fmt.Errorf("username and password not provided")
+		}
+		// BasicAuth implements transport.AuthMethod interface
+		r.auth = &plumbingHTTP.BasicAuth{
+			Username: username,
+			Password: password,
+		}
+		return nil
 
-// Clone runs git clone with depth 1
-func (g *git) Clone(branch string) error {
-	if branch == "" {
-		return g.git("clone", "--depth=1", "-n", "--", g.URL, g.Directory)
-	}
-	return g.git("clone", "--depth=1", "-n", "--branch="+branch, "--", g.URL, g.Directory)
-}
+	case corev1.SecretTypeSSHAuth: // SSH AUTHENTICATION
+		// Make signer from private keys
+		pemBytes := r.secret.Data[corev1.SSHAuthPrivateKey]
+		signer, err := ssh.ParsePrivateKey(pemBytes)
+		if err != nil {
+			return fmt.Errorf("failed to parse ssh private key: %w", err)
+		}
 
-// Update updates git repo if remote sha has changed
-func (g *git) Update(branch string) (string, error) {
-	if err := g.clone(branch); err != nil {
-		return "", nil
-	}
+		// PublicKeys implements transport.AuthMethod interface
+		r.auth = &plumbingSSH.PublicKeys{
+			User:   "git",
+			Signer: signer,
+		}
 
-	if err := g.reset("HEAD"); err != nil {
-		return "", err
-	}
+		// Create temporary known_hosts file
+		hostsBytes := r.secret.Data["known_hosts"]
+		if len(hostsBytes) > 0 {
+			f, err := os.CreateTemp("", "known_hosts")
+			if err != nil {
+				return fmt.Errorf("failed to create temporar known_hosts file: %w", err)
+			}
+			// Write received knownhosts from kubernetes secret
+			_, err = f.Write(hostsBytes)
+			if err != nil {
+				return fmt.Errorf("failed to write knonw_hosts file: %w", err)
+			}
+			// Create callback from recently created temporary file
+			hostKeyCB, err := plumbingSSH.NewKnownHostsCallback(f.Name())
+			if err != nil {
+				return err
+			}
 
-	commit, err := g.currentCommit()
-	if err != nil {
-		return commit, err
-	}
+			r.auth = &plumbingSSH.PublicKeys{
+				User:                  "git",
+				Signer:                signer,
+				HostKeyCallbackHelper: plumbingSSH.HostKeyCallbackHelper{HostKeyCallback: hostKeyCB},
+			}
 
-	if changed, err := g.remoteSHAChanged(branch, commit); err != nil || !changed {
-		return commit, err
-	}
+			// Clean known_hosts file after setting up the callback
+			err = f.Close()
+			if err != nil {
+				return fmt.Errorf("failed to close known hosts file: %w", err)
+			}
+			err = os.Remove(f.Name())
```

Comment:
If we are deleting this file and then the callback by plumbingSSH will not find the file and it will fail right ? Did you test this out ?

Reply:
Did test, the contents of the file are saved in memory here:
`			hostKeyCB, err := plumbingSSH.NewKnownHostsCallback(f.Name())
			if err != nil {
				return err
			}`
			
There is a unit-test that tests this at:
At git_test.go on line 100:
`		{"#3.2 SSH && Known Hosts Secret: Success", &corev1.Secret{
			Type: corev1.SecretTypeSSHAuth,
			Data: secretSSHKnowHostsAuth,
		}, "cattle-test-namespace", "charts-test", "https://somerandom.git", false, []byte{}, nil},`
		
Also, the old code had the same behavior, I just adapted it to use go-git.

Reply:
I have implemented a fix, please resolve the conversation